sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,605 @@
+  /* ... */
+};
+```
+
+(In reality, we can use partial specialization tricks to decorate `atomic_ref`
+with the attribute only when the underlying type is 64-bits.  See ["Appendix:
+Adding an attribute to 8-byte `atomic_ref`"][4].)
+
+[4]: <#appendix-adding-an-attribute-to-8-byte-atomic_ref>
+
+In cases where the optional feature corresponds to a function, we can decorate
+the function's declaration with the attribute like so (demonstrating a
+hypothetical AMX multiplication extension):
+
+```
+[[sycl::requires(has(aspect::ext_intel_amx))]]
+void amx_multiply();
+```
+
+These attributes provide an explicit mapping between each optional kernel
+feature and its associated aspect.
+
+Unfortunately, the fundamental type `double` is also an optional kernel
+feature.  Since there is no type alias for `double`, there is no convenient
+place to add an attribute.  Instead, the front-end device compiler must behave
+as though there was an implicit `[[sycl::requires(has(aspect::fp64))]]`
+attribute for any device code that uses the `double` type.
+
+Note that the usage of `[[sycl::requires()]]` is slightly expanded here beyond
+the specified usage in the SYCL 2020 specification because we allow the
+attribute also on type alias declarations and class declarations.  If a device
+function does any of the following with a type alias or class that was so
+decorated, the function is assumed to "use the aspects" listed in the
+attribute:
+
+* Declares a variable of that type.
+* Has a formal parameter declared with that type.
+* Returns that type.
+
+This also includes any qualified version of the type.
+
+**TODO**: This language is not very precise.  The intent is to include most
+uses of the type, except for cases like `sizeof(T)` or `decltype(T)`.  Help
+appreciated on tightening the wording here.
+
+**TODO**: Would it be better to use a different attribute name when decorating
+types, rather than expanding the meaning of `[[sycl::requires()]]`?  If we did
+this, the new attribute would become an internal DPC++ implementation detail;
+we would not add it to the SYCL specification.
+
+### Implementing diagnostics in the DPC++ front-end
+
+As noted above, the front-end device compiler must behave as though there is an
+implicit `[[sycl::requires(has(aspect::fp64))]]` attribute on any use of the
+`double` type in device code.
+
+Aside from this, the front-end compiler can implement the required diagnostics
+purely from the C++ attributes that have been added to the DPC++ headers.
+There is no need for the front-end compiler to know which device features are
+optional.
+
+When the front-end compiler sees a kernel or device function that is decorated
+with `[[sycl::requires()]]`, it forms the set of allowed aspects for that
+kernel or device function using aspects listed in the attribute.  Let's call
+this the `Allowed` set.  The front-end then computes the static call tree of
+that kernel or device function and forms the union of all aspects in any
```

Comment:
The FE generally does not form call trees (static or otherwise) -- it processes the attribute at the declaration or statement as it is seen and diagnoses what's there. This makes me worry we're designing another feature that expects heroics from the FE rather than making use of the analyzer or analysis passes.

(Note, we already have had to implement some of those heroics for other things like making sure you don't call undefined functions unless they're marked `SYCL_EXTERNAL`. That's not permission to add more instances of the same, though -- my hope is to eventually see those torn out of the FE, likely when we go to upstream to community, specifically because they're implemented in the wrong part of the compiler.)

Reply:
I'm marking this conversation resolved because I copied the comment to another thread and responded to it there.